### PR TITLE
feat(user apis): add pagination to list users api

### PR DESCRIPTION
**Pull request by maghirardelli**
Issue #, if available: GALI-2493

Description of changes:
Add pagination to List Users API

Checklist:

* [x] Have you successfully deployed to an AWS account with your changes?
* [x] Have you written new tests for your core changes, as applicable?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: solutions/swb-reference/RSW.openapi.json [hunk 1]
```diff
@@ -2299,6 +2299,22 @@
         "description": "List Users",
         "operationId": "listUsers",
         "parameters": [
+          {
+            "name": "paginationToken",
+            "in": "query",
+            "schema": {
+              "type": "string",
+              "example": ""
```
Comment (htztomic) [inline solutions/swb-reference/RSW.openapi.json:2307]: do we need an example here?

File: solutions/swb-reference/integration-tests/tests/isolated/users/listUsers.test.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  beforeAll(async () => {
+    adminSession = await setup.getDefaultAdminSession();
+  });
+
+  afterAll(async () => {
+    await setup.cleanup();
+  });
+
+  describe('with invalid paginationToken', () => {
```
Comment (htztomic) [inline solutions/swb-reference/integration-tests/tests/isolated/users/listUsers.test.ts:27]: should include a test to throw invalid parameters to ensure we are validating request inputs

File: solutions/swb-reference/integration-tests/tests/isolated/users/listUsers.test.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    const pagToken = '1';
+    const queryParams = { paginationToken: pagToken };
```
Comment (htztomic) [inline solutions/swb-reference/integration-tests/tests/isolated/users/listUsers.test.ts:29]:
```suggestion
    const queryParams = { paginationToken: pageToken };
```

File: workbench-core/accounts/src/services/projectService.ts [hunk 4]
```diff
@@ -353,10 +354,13 @@ export default class ProjectService {
   ): Promise<PaginatedResponse<Project>> {
     if (paginationToken) {
       const manualExclusiveStartKey = fromPaginationToken(paginationToken);
+      if (typeof manualExclusiveStartKey === 'string') {
```
[the remark this answers is not in the document]
Comment (htztomic) [reply workbench-core/accounts/src/services/projectService.ts:357]: I see, it includes the changes to fromPaginationToken, what would be the scenario to parse a pagination token and not receive an object?

File: workbench-core/user-management/config/jest.config.json [hunk 5]
```diff
@@ -5,8 +5,8 @@
     "global": {
       "branches": 100,
       "functions": 100,
-      "lines": 100,
-      "statements": 100
+      "lines": 99,
```
Comment (htztomic) [inline workbench-core/user-management/config/jest.config.json:8]: reason why we're lowering coverage?

File: workbench-core/user-management/src/plugins/cognitoUserManagementPlugin.test.ts [hunk 6]
```diff
@@ -777,52 +777,132 @@ describe('CognitoUserManagementPlugin tests', () => {
...
+        ]
+      });
+      cognitoMock
+        .on(AdminListGroupsForUserCommand)
+        .resolves({ Groups: roles.map((role) => ({ GroupName: role })) });
+
+      const users = await plugin.listUsers({
+        pageSize: 1,
+        paginationToken:
+          'IkNBSVNxQUlJQVJLQkFnZ0RFdndCQUgwVkljdU9TWFc1TWhieDliN2xZZWRQK3BQVzJ2NzRKbXF6YysySzVycUFleUpBYmlJNklsQmhaMmx1WVhScGIyNURiMjUwYVc1MVlYUnBiMjVFVkU4aUxDSnVaWGgwUzJWNUlqb2lRVUZCUVVGQlFVRkRUakJNUVZGRlFsbFhaV1JyWWpCUVUzQnBVVU40Vm5CV1oxZEtTWGxpSzBWUk5VMUhWVE5oYVRaRGFsRjROR1pWZGpWc1ltMVpOMDVVU1RCT2VsVjZUMGRGZEUxVVFtdFplVEF3VDFSRk0weFVaekZPVkZWMFRVZEthazlYU20xWlYxSnFUMGRLYTA5M1BUMGlMQ0p3WVdkcGJtRjBhVzl1UkdWd2RHZ2lPakVzSW5CeVpYWnBiM1Z6VW1WeGRXVnpkRlJwYldVaU9qRTJPRFV3TnpReE5qSXdOelI5R2lDRnpmNlU4OExHVFlSVU9wdWxCRFQ4Qlh6ZGkyVk55MDVObFVPY2Irci9qQT09Ig=='
```
Comment (htztomic) [inline workbench-core/user-management/src/plugins/cognitoUserManagementPlugin.test.ts:871]: nit: move this to a global variable given the size of it and it is reused

File: workbench-core/user-management/src/plugins/cognitoUserManagementPlugin.test.ts [hunk 7]
```diff
@@ -777,52 +777,132 @@ describe('CognitoUserManagementPlugin tests', () => {
...
+      });
+      cognitoMock
+        .on(AdminListGroupsForUserCommand)
+        .resolves({ Groups: roles.map((role) => ({ GroupName: role })) });
+
+      const users = await plugin.listUsers({ pageSize: 1 });
+
+      expect(users.data.length).toBe(1);
+      expect(users.paginationToken).toBeDefined();
```
Comment (htztomic) [inline workbench-core/user-management/src/plugins/cognitoUserManagementPlugin.test.ts:823]: nit: we should verify that the paginationToken is strictEqual to the value you mocked above

File: solutions/swb-reference/RSW.openapi.json (hunk 1 above)
Comment (maghirardelli) [reply solutions/swb-reference/RSW.openapi.json:2307]: no this is how the other APIs have it in this file

File: solutions/swb-reference/integration-tests/tests/isolated/users/listUsers.test.ts (hunk 2 above)
Comment (maghirardelli) [reply solutions/swb-reference/integration-tests/tests/isolated/users/listUsers.test.ts:27]: Got it! Added a few more to validate `pageSize` and `paginationToken`

File: solutions/swb-reference/integration-tests/tests/isolated/users/listUsers.test.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  beforeAll(async () => {
+    adminSession = await setup.getDefaultAdminSession();
+  });
+
+  afterAll(async () => {
+    await setup.cleanup();
+  });
+
+  describe('with invalid paginationToken', () => {
+    const pagToken = '1';
```
[the remark this answers is not in the document]
Comment (maghirardelli) [reply solutions/swb-reference/integration-tests/tests/isolated/users/listUsers.test.ts:28]: `pag` is short for pagination where `page` is not short for pagination

neither is wrong though! since I am committing some stuff I will go ahead and change `pagToken` to `paginationToken` for simplicity's sake

File: workbench-core/accounts/src/services/projectService.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (maghirardelli) [reply workbench-core/accounts/src/services/projectService.ts:357]: Cognito's pagination token it returns is a string and not an object which is the use case at hand that made me add the type scope to that method!

File: workbench-core/user-management/config/jest.config.json (hunk 5 above)
Comment (maghirardelli) [reply workbench-core/user-management/config/jest.config.json:8]: I couldn't figure out what line was missing from coverage--do you know what command would get me this info?

End of the day I didn't want to be blocked on 1% code cov but if you can help me figure out which line needs to be covered, I can write the test!

File: workbench-core/user-management/src/plugins/cognitoUserManagementPlugin.test.ts (hunk 7 above)
Comment (maghirardelli) [reply workbench-core/user-management/src/plugins/cognitoUserManagementPlugin.test.ts:823]: The paginationToken returned won't equal the pagination token mocked above because we base64 encode the one returned by Cognito within the method (since the pagination token returned by Cognito can have API-URL illegal symbols).

File: solutions/swb-app/src/userRoutes.ts [hunk 9]
```diff
@@ -52,8 +54,23 @@ export function setUpUserRoutes(router: Router, userService: UserManagementServi
...
+      });
+      try {
+        const users = await userService.listUsers(validatedRequest);
+        res.status(200).json({ users });
+      } catch (e) {
+        if (Boom.isBoom(e)) {
+          throw e;
+        }
+
+        if (isInvalidPaginationTokenError(e)) {
```
Comment (nguyen102) [inline solutions/swb-app/src/userRoutes.ts:68]: Nitpick: It's outside the scope of this PR, but since we use this code segment repeatedly (line 68 to 70), we should move it to `errorHandler.ts`. Then we add that middleware in `generateRouter`. That way our code will by more DRY and we won't have to add line 68 to 70 for each paginated API.

File: workbench-core/user-management/config/jest.config.json (hunk 5 above)
Comment (nguyen102) [reply workbench-core/user-management/config/jest.config.json:8]: Add this line to `package.json` of this package
```
"jest": "jest",
```
Run `rushx jest --coverage`

Coverage report will be generated in that package in the `coverage` folder

File: workbench-core/user-management/src/plugins/cognitoUserManagementPlugin.ts [hunk 10]
```diff
@@ -389,21 +397,28 @@ export class CognitoUserManagementPlugin implements UserManagementPlugin {
...
+    let parsedPaginationToken;
+    if (request.paginationToken) {
+      parsedPaginationToken = fromPaginationToken(request.paginationToken);
+    }
     try {
       const response = await this._aws.clients.cognito.listUsers({
-        UserPoolId: this._userPoolId
+        UserPoolId: this._userPoolId,
+        Limit: request.pageSize,
+        PaginationToken: typeof parsedPaginationToken === 'string' ? parsedPaginationToken : undefined
```
Comment (nguyen102) [inline workbench-core/user-management/src/plugins/cognitoUserManagementPlugin.ts:417]: Do we need to do this check? When we call `fromPaginationToken`, doesn't it return a string?

Comment (nguyen102) [reply workbench-core/user-management/src/plugins/cognitoUserManagementPlugin.ts:417]: Btw, are we using `fromPaginationToken` correctly? The [method](https://github.com/aws-solutions/research-service-workbench-on-aws/blob/2645d12a07ae9b9ba185e234713c55639e96291d/workbench-core/base/src/utilities/paginationHelper.ts#L46) returns a dictionary, but we're expecting a string?

File: workbench-core/user-management/src/plugins/cognitoUserManagementPlugin.ts [hunk 11]
```diff
@@ -421,7 +436,13 @@ export class CognitoUserManagementPlugin implements UserManagementPlugin {
         })
       );
 
-      return users.filter((user) => user.id);
+      const data = users.filter((user) => user.id);
```
Comment (nguyen102) [inline workbench-core/user-management/src/plugins/cognitoUserManagementPlugin.ts:439]: What does this filter method do? Is it returning all users that have an `id`? Shouldn't all users have an `id`?

**Review by nguyen102 — COMMENTED** [review]
Do we need to update our multi step users test? I see we have isolated  negative tests. Do we have positive tests?

File: workbench-core/user-management/src/plugins/cognitoUserManagementPlugin.ts (hunk 11 above)
Comment (maghirardelli) [reply workbench-core/user-management/src/plugins/cognitoUserManagementPlugin.ts:439]: It may be for type safe-ness. This was MAF code so I don't have the full context but there are tests to cover not returning users without IDs

**Review by nguyen102 — APPROVED** [review]
